src/EditorFeatures/Core/Extensibility/NavigationBar/AbstractEditorNavigationBarItemService.cs

```diff
@@ -49,20 +57,22 @@ public void NavigateToItem(Document document, NavigationBarItem item, ITextView
                 return;
             }
 
-            var navigationPoint = this.GetSymbolItemNavigationPoint(document, item, cancellationToken);
-
+            var navigationPoint = await this.GetSymbolItemNavigationPointAsync(document, item, cancellationToken).ConfigureAwait(false);
             if (navigationPoint.HasValue)
             {
-                NavigateToVirtualTreePoint(document.Project.Solution, navigationPoint.Value, cancellationToken);
+                await NavigateToVirtualTreePointAsync(document.Project.Solution, navigationPoint.Value, cancellationToken).ConfigureAwait(false);
             }
         }
 
-        protected static void NavigateToVirtualTreePoint(Solution solution, VirtualTreePoint navigationPoint, CancellationToken cancellationToken)
+        protected async Task NavigateToVirtualTreePointAsync(Solution solution, VirtualTreePoint navigationPoint, CancellationToken cancellationToken)
         {
             var documentToNavigate = solution.GetRequiredDocument(navigationPoint.Tree);
             var workspace = solution.Workspace;
             var navigationService = workspace.Services.GetRequiredService<IDocumentNavigationService>();
 
+            // Have to move back to UI thread in order to navigate.
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
simple to switch to teh UI thread when we need it when doign some actual final part of the selection operation.  previously we had to do everything on the UI thread, in a completely blocking fashion to support this.

Reply:
note: i'd still like a future PR where the services just return data as to where they want to navigate to, and the controller takes over.  however, it was a bit contorted given that some services generate items.   so i left that out of this pr for now.